### PR TITLE
PAD bug fix. Avoiding unnecessary marshalling

### PR DESCRIPTION
**Pull request by anderson-joyle**
Issue #790.
Side comments/questions:

- Note that PAD temporary code is part of this PR. This is just to proof the changes have taken action over their test cases.
- Still taking on additional test cases to include. Any suggestions?

### CODE REVIEW

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

File: src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperTableValue.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+namespace Microsoft.PowerFx.Types
+{
+    // Project the correct compile-time type onto the runtime value. 
+    // Important for union / intersection types, such as Table() or If(). For example:
+    //    First(Table({a:1},{b:2})), result is a record with both fields a and b. 
+    internal class CompileTimeTypeWrapperTableValue : CollectionTableValue<DValue<RecordValue>>
+    {
+        private readonly TableValue _inner;
+
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Core/Public/Values/CompileTimeTypeWrapperTableValue.cs:18]: actually, I wonder if we can remove this field altogether?

You're using inheritence to forward, which looks like it solves this. 
In fact, maybe this means we  can even remove it from CompileTimeTypeWrapperRecordValue. 

This is the classic "ineheritence vs. encapsulation" design pattern debate. :)

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Comment by BruceHaley** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Connectors.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Core.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Interpreter.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.LanguageServerProtocol.dll**
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.PowerFx.Transport.Attributes.dll**

**Pull request merged**